src/compiler/scala/tools/nsc/transform/Constructors.scala

```diff
@@ -659,66 +628,106 @@ abstract class Constructors extends Statics with Transform with ast.TreeDSL {
 
             initPhase += mkAssign(statSym, initializingRhs)
           }
+          !const // needs field unless it's a constant
+        }
 
-          defBuf += deriveValDef(stat)(_ => EmptyTree)
-
-        // all other statements go into the constructor
-        case _ => constrStatBuf += intoConstructor(impl.symbol, stat)
+        stat match {
+          // recurse on class definition, store in defBuf
+          case _: ClassDef => defBuf += new ConstructorTransformer(unit).transform(stat)
+
+          // Triage methods -- they all end up in the template --
+          // regular ones go to `defBuf`, secondary contructors go to `auxConstructorBuf`.
+          // The primary constructor is dealt with separately (we're massaging it here).
+          case _: DefDef if statSym.isPrimaryConstructor => ()
+          case _: DefDef if statSym.isConstructor => auxConstructorBuf += stat
+          case _: DefDef => defBuf += stat
+
+          // If a val needs a field, an empty valdef goes into the template.
+          // Except for lazy and ConstantTyped vals, the field is initialized by an assignment in:
+          //   - the class initializer (static),
+          //   - the constructor, before the super call (early initialized or a parameter accessor),
+          //   - the constructor, after the super call (regular val).
+          case ValDef(mods, _, _, rhs) =>
+            val memoized = moveEffectToCtor(mods, rhs)
+
+            if (memoized) defBuf += deriveValDef(stat)(_ => EmptyTree)
```

Comment:
This is a change of behavior, isn't it? Previously a constant type `ValDef` would have a (uninitialized and never read) field, and it doesn't have one anymore. 

I definitely approve of the change, but it would probably be good to mention this change in the commit message, and to have a test case that asserts we don't emit spurious fields anymore.